ksqldb-engine/src/main/java/io/confluent/ksql/query/QueryBuilder.java

```diff
@@ -521,43 +521,44 @@ public NamedTopology getNamedTopology(final SharedKafkaStreamsRuntime sharedRunt
   }
 
   private SharedKafkaStreamsRuntime getKafkaStreamsInstance(
+          final String applicationId,
           final Set<SourceName> sources,
-          final QueryId queryID,
+          final QueryId queryId,
           final MetricCollectors metricCollectors) {
     for (final SharedKafkaStreamsRuntime sharedKafkaStreamsRuntime : streams) {
-      if (sharedKafkaStreamsRuntime.getSources().stream().noneMatch(sources::contains)
-          || sharedKafkaStreamsRuntime.getQueries().contains(queryID)) {
-        sharedKafkaStreamsRuntime.markSources(queryID, sources);
+      if (sharedKafkaStreamsRuntime.getApplicationId().equals(applicationId)
+          || (sharedKafkaStreamsRuntime.getApplicationId().equals(applicationId + "-validation")
+          && !real)) {
+        sharedKafkaStreamsRuntime.markSources(queryId, sources);
```

Comment:
why do we need to bookkeep sources in `SharedKafkaStreamsRuntime` now?

Reply:
We might not now. I would like to keep doing it for now because it should help with the restore. If I find a better way when doing that I can remove it then